lucetc/src/function.rs

```diff
@@ -78,6 +84,138 @@ impl<'a> FuncInfo<'a> {
                 fref
             })
     }
+
+    fn update_instruction_count_instrumentation(
+        &mut self,
+        op: &Operator,
+        builder: &mut FunctionBuilder,
+    ) -> WasmResult<()> {
+        // So the operation counting works like this:
+        // record a stack corresponding with the stack of control flow in the wasm function.
+        // for non-control-flow-affecting instructions, increment the top of the stack.
+        // for control-flow-affecting operations (If, Else, Unreachable, Call, End, Return, Block,
+        // Loop, BrIf, CallIndirect), update the wasm instruction counter and:
+        // * if the operation introduces a new scope (If, Block, Loop), push a new 0 on the
+        // stack corresponding with that frame.
+        // * if the operation does not introduce a new scope (Else, Call, CallIndirect, BrIf),
+        // reset the top of stack to 0
+        // * if the operation completes a scope (End), pop the top of the stack and reset the new
+        // top of stack to 0
+        // * this leaves no special behavior for Unreachable and Return. This is acceptable as they
+        // are always followed by an End and are either about to trap, or return from a function.
+        // * Unreachable is either the end of VM execution and we are off by one instruction, or,
+        // is about to dispatch to an exception handler, which we should account for out of band
+        // anyway (exception dispatch is much more expensive than a single wasm op)
+        // * Return corresponds to exactly one function call, so we can count it by resetting the
+        // stack to 1 at return of a function.
+
+        fn flush_counter(environ: &mut FuncInfo, builder: &mut FunctionBuilder) {
+            if environ.op_offsets.last() == Some(&0) {
+                return;
+            }
+            let instr_count_offset: ir::immediates::Offset32 =
+                (-(std::mem::size_of::<InstanceRuntimeData>() as i32)
+                    + offset_of!(InstanceRuntimeData, instruction_count) as i32)
+                    .into();
+            let vmctx_gv = environ.get_vmctx(builder.func);
+            let addr = builder.ins().global_value(environ.pointer_type(), vmctx_gv);
```

Comment:
This ends up inserting a clif sequence along the lines of
```
let instruction_count_ptr: &mut u64 = vmctx.instruction_count;
let mut instruction_count: u64 = *instruction_count_ptr;
instruction_count += 6;
*instruction_count_ptr = instruction_count;
```
which is a bit verbose, but should be straightforward for Cranelift to optimize down a bit.

this _actually_ turns into sequences like
```
mov rdx, qword [rbp + 8] ; load spilled vmctx
mov rdx, qword [rdx - 8] ; load instruction_count
add rdx, 6 ; update count
mov rbx, qword [rbp + 8]; load spilled vmctx
mov qword [rbx - 8], rdx ; store instruction count
```
which is again a bit worse.

To improve this we'll have to look forward to live-range splitting, count instructions using a pinned register, or use a different strategy to count (which is subject to the same live-range splitting needs).

live-range splitting helps because the vmctx loads shouldn't happen anymore (right now they happen because cranelift must spill `rdi`, and afterward the loads happen from the spilled location), and other long-lived variables will probably not spill in the same way without significant register pressure. That means in a function-local context we can probably store `instruction_count_ptr` in a local, and this should all reduce to `add [instruction_count_ptr], 6`.

The other option, counting into a pinned register, would probably be faster, but seems like more work to implement, and deeper changes Cranelift-side. I'd like to measure the impact of this instrumentation after live-range splitting has a chance to help, so we can figure out if we want to do that instead.

Reply:
Live-range splitting should also help the register allocator do better for general-purpose code, right? Not just for our instruction counting special case

Reply:
> This ends up inserting a clif sequence along the lines of
> 
> ```
> let instruction_count_ptr: &mut u64 = vmctx.instruction_count;
> let mut instruction_count: u64 = *instruction_count_ptr;
> instruction_count += 6;
> *instruction_count_ptr = instruction_count;
> ```

It would be nice to have this as a comment

Reply:
Yup, live-range splitting will help generally, but because of how instruction counting is implemented here it'll be enough of a change I think it's worth noting